src/python/pants/util/objects.py

```diff
@@ -47,8 +47,8 @@ def datatype(field_decls, superclass_name=None, **kwargs):
 
   class DataType(namedtuple_cls):
     @classmethod
-    def make_type_error(cls, msg):
-      return TypeCheckError(cls.__name__, msg)
+    def make_type_error(cls, msg, *args, **kwargs):
```

Comment:
This seems like a weird change. Why are we now accepting arguments that we'll ignore?

Reply:
The idea here is that you can pass in e.g. another exception instance as the next positional argument, so that its stacktrace can be preserved. `*args` and `**kwargs` are eventually forwarded to the `Exception` class constructor, unless I'm mistaken, so I don't see how we're ignoring these? I may be missing something quite obvious.

Reply:
Aah, sorry, I was totally misreading the `TypeCheckError` constructor!

Reply:
Part of that might be because I don't know how normal humans format python code, part of that might be because this chaining of exceptions is currently a little ad-hoc/awkward. See the [TODO in one of the other exception classes in that file](https://github.com/cosmicexplorer/pants/blob/33d0edb1a76bff01a165243b572a510d25cca324/src/python/pants/util/objects.py#L143) -- there could be a method to make this pattern more concise.